llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,435 @@
+          }
+
+        } else {
+          // keep track of already visted physical and virtual Regs
+          UsedPhyRegs.push_back(VRM->getPhys(Reg));
+          UsedVirtRegs.push_back(Register::virtReg2Index(Reg));
+        }
+      }
+    }
+    // dump Reassignment
+    LLVM_DEBUG(dbgs() << "WAW Replacement Overview for bb." << MBB->getNumber();
+               dbgs() << "\n");
+    for (std::pair<MCRegister, MCRegister> &RegPair : ReplacementRegs) {
+      LLVM_DEBUG(dbgs() << printReg(RegPair.first, TRI, 0, MRI) << " -> "
+                        << printReg(RegPair.second, TRI, 0, MRI) << "\n");
+    }
+    LLVM_DEBUG(dbgs() << "\n");
+  }
+  return Modified;
+}
+
+bool AIEWawRegRewriter::isWAWCandidate(
+    const Register &Reg, const MachineInstr &MI,
+    const SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+    const SmallVector<MCPhysReg, 16> &UsedVirtRegs) {
+  const bool FoundPhyReg =
+      UsedPhyRegs.end() != find(UsedPhyRegs, VRM->getPhys(Reg));
+  const bool FoundVirtReg =
+      UsedVirtRegs.end() == find(UsedVirtRegs, Register::virtReg2Index(Reg));
+
+  return MRI->getRegClass(Reg) && FoundPhyReg && FoundVirtReg && !MI.isCopy();
+}
+
+SmallSet<MCPhysReg, 16>
+AIEWawRegRewriter::getMBBForbiddenReplacements(const MachineBasicBlock *MBB) {
+  SmallSet<MCPhysReg, 16> BlockedPhysRegs;
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+      addAliasRegs(BlockedPhysRegs, getPhysReg(Op));
+    }
+  }
+
+  LLVM_DEBUG(dbgs() << "Forbidden Physical Register of the bb."
+                    << MBB->getNumber() << "\n");
+  for (auto &Reg : BlockedPhysRegs) {
+    LLVM_DEBUG(dbgs() << printReg(Reg, TRI, 0, MRI) << " ");
+  }
+
+  return BlockedPhysRegs;
+}
+
+bool AIEWawRegRewriter::replaceReg(const Register Reg,
+                                   SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs) {
+  LLVM_DEBUG(dbgs() << " WAW RegRewriter: Register to replace "
+                    << TRI->getName(VRM->getPhys(Reg)) << "\n");
+  SmallVector<MCPhysReg, 16> Replacements =
+      getPriorityReplacementRegs(Reg, ForbiddenPhysRegs);
+
+  for (const MCPhysReg &PhysReg : Replacements) {
+
+    const LiveInterval &LI = LIS->getInterval(Reg);
+    LiveRegMatrix::InterferenceKind IK = LRM->checkInterference(LI, PhysReg);
+
+    if (IK == LiveRegMatrix::IK_Free) {
+      LLVM_DEBUG(dbgs() << "     WAW Replacement: Virtual Register "
+                        << Register::virtReg2Index(Reg) << " will replace "
+                        << printReg(VRM->getPhys(Reg), TRI, 0, MRI) << " with "
+                        << printReg(PhysReg, TRI, 0, MRI) << '\n');
+      LRM->unassign(LI);
+      LRM->assign(LI, PhysReg);
+      addAliasRegs(ForbiddenPhysRegs, PhysReg);
+      return true;
+    }
+  }
+  return false;
+}
+
+SmallVector<MCPhysReg, 16> AIEWawRegRewriter::getPriorityReplacementRegs(
+    const Register Reg, const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs) {
```

Comment:
Nit: Add something like `assert(Reg.isVirtual())` to verify the precondition of this function

Reply:
added the check, but the getPriorityReplacementRegs function is private and the only call it has, only allows Reg to be virtual. So the check should be unnecessary

Reply:
I know it's unnecessary, that's why it's an assert :) The point is to verify the pre-conditions of our functions, those asserts should never trigger, but you never know how the code will evolve in the future

Reply:
added an assert